### PR TITLE
feat: Two Factor SMS token parser and copy to clipboard via notifications

### PR DESCRIPTION
**Pull request by dutscher**
#### Pre-flight Checklist

Please ensure you've completed all of the following.

- [X] I have read the [Contributing Guidelines](https://github.com/ferdium/ferdium-app/blob/HEAD/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/ferdium/ferdium-app/blob/HEAD/CODE_OF_CONDUCT.md) that this project adheres to.

#### Description of Change

I have the daily trouble with two factor tokens they came via sms on my mobile phone. My ferdium is connected via android messages and the app shows every sms via ferdium notifications on my desktop. I wrote a little helper which parse the token out of the notification body and copy the found token into clipboard. Now i can press CTRL + V instead of look onto the notification and tip every digit into my input.

#### Motivation and Context

See description :D My motivation is my daily pain with tipping every digit or copy the tokens with accidentally copied whitespaces.

#### Checklist

- [X] My pull request is properly named
- [X] The changes respect the code style of the project (`pnpm prepare-code`)
- [X] `pnpm test` passes
- [X] I tested/previewed my changes locally

#### Release Notes

Two Factor SMS Token can now automatically parsed from notification (Ex.: android messages) and its copied to clipboard.

Second try. Now in a clean pull request with settings.

Thanks in advance

### CODE REVIEW

**Comment by dutscher** [comment]
This is a clean pull request of the first pr created of mine https://github.com/ferdium/ferdium-app/pull/1555
Now i try to test the notifications again on my windows 11.
Cheers

File: src/@types/stores.types.ts
Comment (vraravam) [inline src/@types/stores.types.ts]: Looking through the code, this seems like a toggle? If so, please differentiate between the toggle and the value via more appropriate names. For eg, the toggle can be `twoFactorAutoCatcherFlag` and the other variable could be without the `Array` suffix.

File: src/components/settings/settings/EditSettingsForm.tsx
Comment (vraravam) [inline src/components/settings/settings/EditSettingsForm.tsx:370]: By this code, it looks like the array will always have a value? But, if I, as a security conscious user, want to turn this off, then shouldn't the array remain empty?

File: src/containers/settings/EditSettingsScreen.tsx
Comment (vraravam) [inline src/containers/settings/EditSettingsScreen.tsx]: typo: separated
also "comma-separated and case-insensitive" - hyphenated

File: src/i18n/locales/en-US.json
Comment (vraravam) [inline src/i18n/locales/en-US.json]: same as above

**Review by vraravam — CHANGES_REQUESTED** [review]

**Comment by vraravam** [comment]
@dutscher  - please do not close old PRs and create duplicate / new ones. The reviewers would like to go through the history of their requests and see how the code has been modified.

Also, please rebase from the parent branch (develop) rather than merging from it. This provides a much cleaner view to the reviewers before approving.
Screenshot 2024-02-10 at 5 30 42 PM

**Comment by dutscher** [comment]
> @dutscher - please do not close old PRs and create duplicate / new ones. The reviewers would like to go through the history of their requests and see how the code has been modified.
> 
> Also, please rebase from the parent branch (develop) rather than merging from it. This provides a much cleaner view to the reviewers before approving. Screenshot 2024-02-10 at 5 30 42 PM

Thanks @vraravam for this hint. Exactly the merge in the old was the problem and that i create my feature on the develop directly.
I wasn't able to get rid off the recipes changes as you can see here https://github.com/ferdium/ferdium-app/pull/1555/files.
The simplest option i saw was to create a new fork. Sorry for my doing.
Now i have the dev merge commit which comes from the simple click tool from gtihub. How i get this commit out of the history?

Edit: i used now the rebase option and et voila the dev merge ist out of the history.
Now lets clarify the open conversations.

Cheers

File: src/@types/stores.types.ts
Comment (dutscher) [reply src/@types/stores.types.ts]: what about `enableTwoFactorAutoCatcher` or `isTwoFactorAutoCatcherEnabled`? 
the other toogle's don't have straight naming concept which we can follow :)
i prefer `isTwoFactorAutoCatcherEnabled` and the array `twoFactorAutoCatcherMatcher`.

File: src/components/settings/settings/EditSettingsForm.tsx
Comment (dutscher) [reply src/components/settings/settings/EditSettingsForm.tsx:370]: 1. yes
2. yes

maybe we get, if the toggle get's activated, then the array get's filled?

File: src/containers/settings/EditSettingsScreen.tsx
Comment (dutscher) [reply src/containers/settings/EditSettingsScreen.tsx]: solved

File: src/i18n/locales/en-US.json
Comment (dutscher) [reply src/i18n/locales/en-US.json]: solved

File: src/@types/stores.types.ts
Comment (vraravam) [reply src/@types/stores.types.ts]: pls use `isTwoFactorAutoCatcherEnabled` - I actually noticed the same in 2 other feature-enablement flags, and have a PR open (waiting for approval) for exactly this reason
the 2nd variable (I dont think the datatype is an array, its more of a comma-separated string), can be `twoFactorAutoCatcherMatcher`

File: src/components/settings/settings/EditSettingsForm.tsx
Comment (vraravam) [reply src/components/settings/settings/EditSettingsForm.tsx:370]: yes, if the toggle gets activated, the value of the string can be non-empty
If the toggle gets deactivated, the value of the string can become empty string

File: src/containers/settings/EditSettingsScreen.tsx
Comment (vraravam) [inline src/containers/settings/EditSettingsScreen.tsx]: hyphenated: `Auto-catch`

File: src/i18n/locales/en-US.json
Comment (vraravam) [inline src/i18n/locales/en-US.json]: hyphenated `Auto-catch`

**Review by vraravam — CHANGES_REQUESTED** [review]

**Comment by vraravam** [comment]
> The simplest option i saw was to create a new fork. Sorry for my doing.

You can use a gui-based tool (I use git-gui) to amend the previous commit, and remove (ie unstage) the `recipes` from that commit. This will then remove it from the PR - once you force-push to the remote.

> Now i have the dev merge commit which comes from the simple click tool from gtihub. How i get this commit out of the history?

Rebasing from the `develop` branch will automatically remove the merge commit - since it will result in a "no-changes" commit, which git intelligently ignores.

File: src/@types/stores.types.ts
Comment (dutscher) [reply src/@types/stores.types.ts]: done

File: src/containers/settings/EditSettingsScreen.tsx
Comment (dutscher) [reply src/containers/settings/EditSettingsScreen.tsx]: done

File: src/i18n/locales/en-US.json
Comment (dutscher) [reply src/i18n/locales/en-US.json]: done

**Comment by dutscher** [comment]
## Default:
image

## Enabled is prefilled:
image

### Console: 
image

## Disabled the store is empty:
image

I'm not able to test the notifivcation.ts part. Windows 11 only accept installed Ferdium notification code. Or i'm wrong?

Cheers

**Review by vraravam — APPROVED** [review]
Haven't tested this, but the review comments are all fixed.

**Comment by dutscher** [comment]
@vraravam on which platform did you work?
@SpecialAro did you know how i can test on win11 the code from notifications.ts?

If i start `debug` i didn't get some `Ferdium:Notifications`.

cheers

**Comment by vraravam** [comment]
im on macos ventura beta running on an m2 (arm) machine

**Comment by SpecialAro** [comment]
> @SpecialAro did you know how i can test on win11 the code from notifications.ts?
> 
> If i start `debug` i didn't get some `Ferdium:Notifications`.
> 
> cheers

Maybe you can build it entirely (run `./scripts/build-windows.ps1`) and run the `Ferdium.exe` that is on the `out` folder. You can then user Ferdium normally with your services and see if this works.

**Comment by dutscher** [comment]
now i know why i didnt get the debug logs.

```
Ferdium:Services Service logged a message: +0ms [Ferdium:Notifications] error TypeError: Cannot read properties of undefined (reading 'stores')
```

comes from

```
const { isTwoFactorAutoCatcherEnabled, twoFactorAutoCatcherMatcher } =
          window['ferdium'].stores.settings.app;
```

how we get that fixed? access to the ferdium store int possbile from notifications.ts

File: src/components/settings/settings/EditSettingsForm.tsx
Comment (SpecialAro) [inline src/components/settings/settings/EditSettingsForm.tsx]: are these changes needed? when the user disables the autocatcher, the matcher still stores (in the config file) the strings. This is intended (so that a user can disable and enable the feature, without having to edit everything if enabled again)

**Review by SpecialAro — CHANGES_REQUESTED** [review]

**Comment by SpecialAro** [comment]
> now i know why i didnt get the debug logs.
> 
> ```
> Ferdium:Services Service logged a message: +0ms [Ferdium:Notifications] error TypeError: Cannot read properties of undefined (reading 'stores')
> ```
> 
> comes from
> 
> ```
> const { isTwoFactorAutoCatcherEnabled, twoFactorAutoCatcherMatcher } =
>           window['ferdium'].stores.settings.app;
> ```
> 
> how we get that fixed? access to the ferdium store int possbile from notifications.ts

I just fixed this by moving the entire logic from `notifications.ts` to the `ServicesStore.ts` (which is the correct place to add this logic to).

Can you please check again if this works?

**Comment by dutscher** [comment]
> > ...
> 
> Can you please check again if this works?

it works. thanks for this change!
image

And here the "+" in the windows notification:
image

File: src/components/settings/settings/EditSettingsForm.tsx
Comment (dutscher) [reply src/components/settings/settings/EditSettingsForm.tsx]: if its get disabled the store is filled with a string which was a security issue for @vraravam

Comment (vraravam) [reply src/components/settings/settings/EditSettingsForm.tsx]: yes, i had requested for this to be reset/empty if the user turned it off. But, if you feel this is better for the user, then I can be convinced.

Comment (SpecialAro) [reply src/components/settings/settings/EditSettingsForm.tsx]: @vraravam I think that we should allow the user to enable/disable without changing the strings. We just have to make sure we don't catch the notifications when the feature is off (which we dont, so that's ok). Nevertheless, if you feel we should make it this way, I would suggest to change it to the default settings, and not to an empty string.

@dutscher would you be able to revert this change? this way we could merge it :D

**Review by SpecialAro — CHANGES_REQUESTED** [review]

File: src/components/settings/settings/EditSettingsForm.tsx
Comment (dutscher) [reply src/components/settings/settings/EditSettingsForm.tsx]: its done and reverted to the first implementation of your code @SpecialAro 
lets merge this :)

cheers and thanks for your support!

**Review by SpecialAro — APPROVED** [review]
LGTM! 😁 Should we merge this @vraravam ?

**Comment by vraravam** [comment]
> LGTM! 😁 Should we merge this @vraravam ?

LGTM!

**Comment by dutscher** [comment]
who is able to merge?

**Comment by vraravam** [comment]
@all-contributors please add @dutscher for code

**Comment by MurzNN** [comment]
@dutscher Thanks for the feature, it's really useful! But can it be configured per service? To parse messages only from the chosen services, not from all at once?

**Comment by dutscher** [comment]
@MurzNN for sure this is possible. Do you have problems with parse all the notifications and your clipboard?
Aswell i thought about a timeout to clear the clipboard, the same way as keypass did.

Cheers

**Comment by MurzNN** [comment]
> @MurzNN for sure this is possible. Do you have problems with parse all the notifications and your clipboard?

I still can't find where this can be configured, could you please add screenshots? And how to debug it if it is not working well? Or maybe even post a short howto about this to the docs here https://github.com/ferdium/ferdium-app/tree/develop/docs ?

**Comment by dutscher** [comment]
with "for sure this is possible" i mean that we need to develop this configure per service.
actual you can only configure in global privacy settings:
image

cheers

**Pull request merged**
